modules/drivers/mongo/src/metabase/driver/mongo/java_driver_wrapper.clj

```diff
@@ -0,0 +1,341 @@
+
+(defn list-indexes
+  ([^MongoDatabase db coll-name]
+   (list-indexes (collection db coll-name)))
+  ([^MongoCollection coll]
+   (into [] (.listIndexes coll))))
+
+(defn do-find [^MongoCollection coll
+               & {:keys [limit skip batch-size sort-criteria] :as _opts}]
+  (->> (cond-> ^FindIterable (.find coll)
+         limit (.limit limit)
+         skip (.skip skip)
+         batch-size (.batchSize (int batch-size))
+         sort-criteria (.sort (to-document sort-criteria)))
+      (mapv #(from-document % true))))
+
+(defn create-index
+  "Create index."
+  [^MongoCollection coll cmd-map]
+  (.createIndex coll (to-document cmd-map)))
+
+(defn insert-one
+  "Insert document into mongo collection."
+  [^MongoCollection coll document-map]
+  (.insertOne coll (to-document document-map)))
+
+;;;; Util
+
+(defn- fqdn?
+  "A very simple way to check if a hostname is fully-qualified:
+   Check if there are two or more periods in the name."
+  [host]
+  (<= 2 (-> host frequencies (get \. 0))))
+
+(defn- validate-db-details! [{:keys [use-conn-uri conn-uri use-srv host] :as _db-details}]
+  (when (and use-srv (not (fqdn? host)))
+    (throw (ex-info (tru "Using DNS SRV requires a FQDN for host")
+                    {:host host})))
+  (when (and use-conn-uri (empty? (-> (ConnectionString. conn-uri) .getDatabase)))
+    (throw (ex-info (tru "No database name specified in URI.")
+                    {:host host}))))
+
+(defn update-ssl-db-details
+  [db-details]
+  (-> db-details
+      (assoc :client-ssl-key (secret/get-secret-string db-details "client-ssl-key"))
+      (dissoc :client-ssl-key-creator-id
+              :client-ssl-key-created-at
+              :client-ssl-key-id
+              :client-ssl-key-source)))
+
+(defn- details-normalized
+  "Return _normalized_ database `:details` for `x`, where `x` could be a database id, database name, database object
+   (as returned eg. by call to `(toucan2.core/select-one :model/Database)`)"
+  [database]
+  (let [db-details
+        (cond
+          (integer? database)             (qp.store/with-metadata-provider database
+                                            (:details (lib.metadata.protocols/database (qp.store/metadata-provider))))
+          (string? database)              {:dbname database}
+          (:dbname (:details database))   (:details database) ; entire Database obj
+          (:dbname database)              database            ; connection details map only
+          (:conn-uri database)            database            ; connection URI has all the parameters
+          (:conn-uri (:details database)) (:details database)
+          :else
+          (throw (Exception. (str "with-mongo-connection failed: bad connection details:"
+                                  (:details database)))))]
+    (validate-db-details! db-details)
+    (update-ssl-db-details db-details)))
+
+;; TODO: ADD explanation from notion doc.
+(defn db-details->connection-string
```

Comment:
While WIP, more context on building connection string can be found in [this notion doc](https://www.notion.so/metabase/Mongo-without-monger-9f43f2c6a1e14be9bb49a82fd5d65eda?pvs=4). Its contents will be moved here, into docstring, later.